{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "from matplotlib import pyplot as plt\n",
    "import os, wget, subprocess"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "pdb_id = '4AKE'\n",
    "local_storage = './data/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "if not os.path.isfile(f'{local_storage}{pdb_id.lower()}.pdb'):\n",
    "    pdb_url=f'https://files.rcsb.org/download/{pdb_id.upper()}.pdb'\n",
    "    pdb_filename = wget.download(pdb_url, out='./data/')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# NOLB"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "n_decoys=100\n",
    "rmsd=1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "run = subprocess.run(['NOLB', \n",
    "                      f'{local_storage}{pdb_id.lower()}.pdb', \n",
    "                      f'-s {n_decoys}',\n",
    "                      f'--rmsd {rmsd}'\n",
    "                     ], \n",
    "                     stdout=subprocess.PIPE, stderr=subprocess.PIPE\n",
    "                    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "*******************************************************************\n",
      "*-----------------------------------------------------------------*\n",
      "*-----------NOLB : a Non-Linear rigid block NMA method------------*\n",
      "*----------Authors: Alexandre Hoffmann & Sergei Grudinin----------*\n",
      "*-Copyright (c): Nano-D team, Inria/CNRS Grenoble, France, 2019.--*\n",
      "*--------------- e-mail: sergei.grudinin@inria.fr ----------------*\n",
      "*---- http://team.inria.fr/nano-d/software/nolb-normal-modes/ ----*\n",
      "*-----------------------------------------------------------------*\n",
      "*******************************************************************\n",
      "Hostname......................................................... : langev2\n",
      "Run at........................................................... : Mon May 18 16:00:56 2020\n",
      "Command-line arguments........................................... : NOLB ./data/4ake.pdb -s 100 --rmsd 1 \n",
      "=====================Reading input PDB file======================\n",
      "Input PDB file................................................... : ./data/4ake.pdb\n",
      "Number of chains read............................................ : 3\n",
      "Number of atoms read............................................. : 3312\n",
      "======================Constructing Hessian=======================\n",
      "Number of disconnected atoms..................................... : 0\n",
      "Number of interactions........................................... : 79232\n",
      "Interaction cutoff distance...................................... : 5 A\n",
      "Memory required for the Hessian matrix........................... : 16.3213 Mb\n",
      "All-atom Hessian size : ......................................... : 9936 x 9936\n",
      "============================Doing NMA============================\n",
      "Memory required for the RTB projection matrix.................... : 0.909668 Mb\n",
      "Reduced Hessian size : .......................................... : 2568 x 2568\n",
      "Number of computed modes......................................... : 10 \n",
      "Null-space size.................................................. : 1 \n",
      "Lowest mode frequency............................................ : 0.000309662 \n",
      "Highest mode frequency........................................... : 0.00110133 \n",
      "Frequencies :\n",
      "            [ 1]................................................. : 0.000309662 \n",
      "            [ 2]................................................. : 0.000422403 \n",
      "            [ 3]................................................. : 0.000436582 \n",
      "            [ 4]................................................. : 0.000547848 \n",
      "            [ 5]................................................. : 0.000582734 \n",
      "            [ 6]................................................. : 0.00068716 \n",
      "            [ 7]................................................. : 0.000736374 \n",
      "            [ 8]................................................. : 0.000976951 \n",
      "            [ 9]................................................. : 0.001047 \n",
      "            [10]................................................. : 0.00110133 \n",
      "===================Writing output Decoy files====================\n",
      "Number of files.................................................. : 100\n",
      "Output RMSD...................................................... : 1 A\n",
      "Using outputfilename............................................. : ./data/4ake_nlb_decoys.pdb\n",
      "=================================================================\n",
      "=========================== Timing : ============================\n",
      "=================================================================\n",
      "Parsing arguments................................................ : 0.000780241 s\n",
      "Reading input PDB file........................................... : 0.00306677 s\n",
      "Constructing Hessian............................................. : 0.0522912 s\n",
      "Doing NMA........................................................ : 0.269663 s\n",
      "Writing output Decoy files....................................... : 0.54134 s\n",
      ".................................................................\n",
      "Total time : .................................................... : 0.867152 s\n",
      "=================================================================\n",
      "\n",
      "Main Reference: Alexandre Hoffmann & Sergei Grudinin. NOLB : Nonlinear rigid block normal mode analysis method. Journal of Chemical Theory and Computation, 2017, 13 (5), pp.2123-2134. DOI: 10.1021/acs.jctc.7b00197.\n",
      "\n",
      "Reference for clustering and decoy generation: Emilie Neveu, Petr Popov, Alexandre Hoffmann, Angelo Migliosi, Xavier Besseron, Gregoire Danoy, Pascal Bouvry, Sergei Grudinin. RapidRMSD: Rapid determination of RMSDs corresponding to motions of flexible molecules. Bioinformatics, Oxford University Press (OUP), 2018. DOI: 10.1093/bioinformatics/bty160.\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(run.stdout.decode('ascii'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "*******************************************************************\r\n",
      "*-----------------------------------------------------------------*\r\n",
      "*-----------NOLB : a \u001b[1;31mN\u001b[0mon-\u001b[1;31mL\u001b[0minear Rigid \u001b[1;31mB\u001b[0mlock NMA method------------*\r\n",
      "*----------Authors: Alexandre Hoffmann & Sergei Grudinin----------*\r\n",
      "*-Copyright (c): Nano-D team, Inria/CNRS Grenoble, France, 2019.--*\r\n",
      "*--------------- e-mail: \u001b[4;34msergei.grudinin@inria.fr\u001b[0m ----------------*\r\n",
      "*---- \u001b[4;35mhttp://team.inria.fr/nano-d/software/nolb-normal-modes/\u001b[0m ----*\r\n",
      "*-----------------------------------------------------------------*\r\n",
      "*******************************************************************\r\n",
      "\r\n",
      "USAGE: \r\n",
      "\r\n",
      "   NOLB  <pdb filename> <pdb reference file> [-o <output name>] [-n <number\r\n",
      "         of modes>] [--mode <unsigned integer>] ...  [-c <cutoff distance>]\r\n",
      "         [-s <number of output frames>] [-a <maximum amplitude>] [-p\r\n",
      "         <potential function>] [-m] [--linear] [--trajectory] [-j] [-x]\r\n",
      "         [--hetatm] [--zdock <Zdock transform filename>] [--hex <Hex\r\n",
      "         transform filename>] [--frames <max number of input trajectory\r\n",
      "         frames>] [--covScaling <regularization of fuzzy interactions>]\r\n",
      "         [--fuzzy] [-dcd <dcd file name>] [-xrd <xrd file name>] [--clust]\r\n",
      "         [--coords <input traj file>] [--weights <input weights file>] [-r\r\n",
      "         <maximum rmsd>] [--dist <RMSD distribution>] [--rand] [--nSteps\r\n",
      "         <number of minimization steps>] [-t <minimization tolerance>]\r\n",
      "         [--blocks <Rigid blocks ids filename>] [--excl <Excluded\r\n",
      "         interactions filename>] [--fixed <Fixed residues filename>]\r\n",
      "         [--oldFormat] [--analyze] [--collectivity] [--nlin] [--nIter\r\n",
      "         <number of additional Hessian matrix computations>] [--maxIter\r\n",
      "         <number of iterations for nonlinear transitions>]\r\n",
      "         [--alignmentCycles <alignment cycles>] [--alignmentThreashold\r\n",
      "         <alignment threashold for outliers in A>] [--blossumThreashold\r\n",
      "         <alignment blossom threasold>] [--nDomains <number of domains>]\r\n",
      "         [--motionType <type of motion tree>] [--domainsPerMode]\r\n",
      "         [--voronota <Voronota tesselation filename>] [-d <type of\r\n",
      "         dynamics>] [-h] [--version] [--log]\r\n",
      "\r\n",
      "\r\n",
      "Where: \r\n",
      "\r\n",
      "   <pdb filename>\r\n",
      "     (required)  Input PDB file, can be a PDB multi-model trajectory file\r\n",
      "\r\n",
      "   <pdb reference file>\r\n",
      "     Reference PDB file. Only to compute the least RMSD motion. Disabled by\r\n",
      "     default.\r\n",
      "\r\n",
      "   -o <output name>,  --output <output name>\r\n",
      "     Output file name. If nothing provided, a default name will be chosen.\r\n",
      "\r\n",
      "   -n <number of modes>,  --modes <number of modes>\r\n",
      "     Number of modes, 10 by default\r\n",
      "\r\n",
      "   --mode <unsigned integer>  (accepted multiple times)\r\n",
      "     list of normal modes used. all by default\r\n",
      "\r\n",
      "   -c <cutoff distance>,  --cutoff <cutoff distance>\r\n",
      "     Interactions cutoff distance in angstroms, 5 by default\r\n",
      "\r\n",
      "   -s <number of output frames>,  --solutions <number of output frames>\r\n",
      "     Number of frames in the output NMA trajectory, 10 by default. Please\r\n",
      "     note that s=0 suppresses the output.\r\n",
      "\r\n",
      "   -a <maximum amplitude>,  --amp <maximum amplitude>\r\n",
      "     Maximum amplitude, 1.0 by default. The absolute amplitude value is\r\n",
      "     chosen automatically according to the system size. This option only\r\n",
      "     adjusts the absolute value.\r\n",
      "\r\n",
      "   -p <potential function>,  --potential <potential function>\r\n",
      "     Potential function, 0 = Elastic Network, 1 = Gaussian Network.\r\n",
      "     Currently disabled.\r\n",
      "\r\n",
      "   -m,  --minimize\r\n",
      "     Minimize obtained structures, off by default.\r\n",
      "\r\n",
      "   --linear\r\n",
      "     Additionally compute the linear modes. Off by default.\r\n",
      "\r\n",
      "   --trajectory\r\n",
      "     Save the output trajectory at equidistant points. Off by default.\r\n",
      "\r\n",
      "   -j,  --json\r\n",
      "     Output into the log file in the JSON format. False by default. Can be\r\n",
      "     combined with the plain text output.\r\n",
      "\r\n",
      "   -x,  --text\r\n",
      "     Output int the log file in the TEXT format. False by default.\r\n",
      "\r\n",
      "   --hetatm\r\n",
      "     Read heteroatoms from the input file. False by default.\r\n",
      "\r\n",
      "   --zdock <Zdock transform filename>\r\n",
      "     Zdock transforms file name. Experimental option.\r\n",
      "\r\n",
      "   --hex <Hex transform filename>\r\n",
      "     Hex transforms file name. Experimental option.\r\n",
      "\r\n",
      "   --frames <max number of input trajectory frames>\r\n",
      "     Max number of trajectory frames to use for trajectory analysis, 100 by\r\n",
      "     default.\r\n",
      "\r\n",
      "   --covScaling <regularization of fuzzy interactions>\r\n",
      "     Regularization of fuzzy interactions, 100 by default.\r\n",
      "\r\n",
      "   --fuzzy\r\n",
      "     Dumping of fuzzy interactions. Off by default.\r\n",
      "\r\n",
      "   -dcd <dcd file name>,  --dcd <dcd file name>\r\n",
      "     DCD file name. Requires corresponding PDB file.\r\n",
      "\r\n",
      "   -xrd <xrd file name>,  --xrd <xrd file name>\r\n",
      "     XRD file name. Requires corresponding PDB file.\r\n",
      "\r\n",
      "   --clust\r\n",
      "     Cluster trajectory frames, off by default.\r\n",
      "\r\n",
      "   --coords <input traj file>\r\n",
      "     Name of the text trajectory coordinate file that should correspond to\r\n",
      "     input pdb. Experimental option.\r\n",
      "\r\n",
      "   --weights <input weights file>\r\n",
      "     Name of the trajectory weights text file, where weights for different\r\n",
      "     frames are provided. Experimental option.\r\n",
      "\r\n",
      "   -r <maximum rmsd>,  --rmsd <maximum rmsd>\r\n",
      "     Maximum rmsd for decoy generation. In case of MD clustering - RMSD\r\n",
      "     threshold.\r\n",
      "\r\n",
      "   --dist <RMSD distribution>\r\n",
      "     Power dependence of RMSD distribution of generated decoys on the input\r\n",
      "     RMSD, 0 for constant RMSD value (default), 1 for linear dependence,\r\n",
      "     0.5 for quadratic dependence, etc.\r\n",
      "\r\n",
      "   --rand\r\n",
      "     Random seed for decoy generation. Off by default.\r\n",
      "\r\n",
      "   --nSteps <number of minimization steps>\r\n",
      "     Number of minimization steps, 500 by default.\r\n",
      "\r\n",
      "   -t <minimization tolerance>,  --tol <minimization tolerance>\r\n",
      "     Minimization tolerance, 0.1 by default\r\n",
      "\r\n",
      "   --blocks <Rigid blocks ids filename>\r\n",
      "     Rigid blocks ids file name.\r\n",
      "\r\n",
      "   --excl <Excluded interactions filename>\r\n",
      "     Excluded interactions file name. Each line of this file lists chain1\r\n",
      "     and the corresponding range of resIDs followed by chain2 and the\r\n",
      "     corresponding range of resIDs\r\n",
      "\r\n",
      "   --fixed <Fixed residues filename>\r\n",
      "     Fixed residues file name. Each line of this file lists chainID and the\r\n",
      "     corresponding range of resIDs\r\n",
      "\r\n",
      "   --oldFormat\r\n",
      "     Old format of the rigid-block and excluded interactions files. Off by\r\n",
      "     default. In the new format, each line specifies a space-separated set\r\n",
      "     of intervals, for example, 'C299:304 B452:458 B488:496 B514:523\r\n",
      "     A184:436'. In the old format, new rigid block starts with a # line,\r\n",
      "     and each interval is written in a new line, for example, \r\n",
      "\r\n",
      "     # new block\r\n",
      "\r\n",
      "     B558:633\r\n",
      "\r\n",
      "     C437:442\r\n",
      "\r\n",
      "     A437\r\n",
      "\r\n",
      "   --analyze\r\n",
      "     Analyze the computed modes. Off by default.\r\n",
      "\r\n",
      "   --collectivity\r\n",
      "     Prints modes collectivity. Off by default.\r",
      "\r\n",
      "\r\n",
      "   --nlin\r\n",
      "     Deterministic nonlinear transitions if the reference structure is\r\n",
      "     given, off by default.\r\n",
      "\r\n",
      "   --nIter <number of additional Hessian matrix computations>\r\n",
      "     Number of additional computations of the Hessian matrix if the\r\n",
      "     reference structure is given, 0 by default. If enabled, the\r\n",
      "     recommended value is 5.\r\n",
      "\r\n",
      "   --maxIter <number of iterations for nonlinear transitions>\r\n",
      "     Number of iterations for nonlinear transitions, given a computed\r\n",
      "     Hessian matrix, 100 by default. It may be advantageous to increase\r\n",
      "     this value in case of very large transitions.\r\n",
      "\r\n",
      "   --alignmentCycles <alignment cycles>\r\n",
      "     Number of additional alignment cycles, 0 by default\r\n",
      "\r\n",
      "   --alignmentThreashold <alignment threashold for outliers in A>\r\n",
      "     Distance threashold for outliers, 2A by default. Please understand\r\n",
      "     what you are doing when choosing this value. Structural transitions\r\n",
      "     bigger than this value will be suppressed!\r\n",
      "\r\n",
      "   --blossumThreashold <alignment blossom threasold>\r\n",
      "     Blossum similarity threashold between amino acids in the alignment, +1\r\n",
      "     by default\r\n",
      "\r\n",
      "   --nDomains <number of domains>\r\n",
      "     Number of domains to output.\r\n",
      "\r\n",
      "   --motionType <type of motion tree>\r\n",
      "     Type of motion tree, 0 - energy tree, 1 - geometry tree. Default - 0.\r\n",
      "\r\n",
      "   --domainsPerMode\r\n",
      "     Also compute domain decomposition per mode, off by default.\r\n",
      "\r\n",
      "   --voronota <Voronota tesselation filename>\r\n",
      "     Voronota tesselation file name.\r\n",
      "\r\n",
      "   -d <type of dynamics>,  --dynamicsType <type of dynamics>\r\n",
      "     Type of dynamics, 0 - equilibrium (default), 1 - nonequilibrium.\r\n",
      "\r\n",
      "   -h,  --help\r\n",
      "     Displays usage information and exits.\r\n",
      "\r\n",
      "   --version\r\n",
      "     Displays version information and exits.\r\n",
      "\r\n",
      "   --log\r\n",
      "     Displays ChangeLog information and exits.\r\n",
      "\r\n"
     ]
    }
   ],
   "source": [
    "! NOLB --help"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
